### PR TITLE
Simplify matrix configuration for CI workflows

### PR DESCRIPTION
**Pull request by sjain-stanford**
Addresses https://github.com/llvm/torch-mlir/issues/1207. 

#### Provisioned jobs:
```
# ubuntu - x86_64 - llvm in-tree     - pytorch binary - build+test    # most used dev flow and fastest signal
# ubuntu - x86_64 - llvm out-of-tree - pytorch source - build+test    # most elaborate build
# macos  - arm64  - llvm in-tree     - pytorch source - build only    # cross compile, can't test arm64
```

#### Main changes
- [x] Spawn macos builds from a separate matrix (in the same workflow). It made sense to do this as they are fairly different from ubuntu (cross compile, use a different cmake configuration). This simplifies the matrix configuration and exclusions quite a bit, and makes the workflow a bit more tractable and maintenance friendly.
- [x] Remove the submodule md5sum step for ccache config. This was [broken](https://github.com/llvm/torch-mlir/runs/7779288734?check_suite_focus=true#step:3:145) for a while now.
- [x] Removes unused matrix options - `os`, `targetarch`, `python-version`, `llvmtype`.
- [x] Address ZSTD [comment](https://github.com/llvm/torch-mlir/pull/1204#discussion_r942349282) on @powderluv's cross compile [PR](https://github.com/llvm/torch-mlir/pull/1204). 

#### Further improvements (to be addressed in follow-on):
* ubuntu-x86_64 out-of-tree integration tests fail ([error](https://github.com/sjain-stanford/torch-mlir/runs/7781264029?check_suite_focus=true)); only run unit tests for now (tests are excluded in current CI too)

#### Passing workflow:
https://github.com/sjain-stanford/torch-mlir/actions/runs/2840676309
image

### CODE REVIEW

**Comment by powderluv** [comment]
This is looking great. The arm64 Pytorch src build should work if you remove this one line https://github.com/llvm/torch-mlir/blob/51bfe25c898e5f7a09294ccc012cfbf1327614c9/build_tools/build_libtorch.sh#L134  
It is trying to uninstall the systemwide torch.

File: .github/workflows/buildAndTest.yml
Comment (powderluv) [inline .github/workflows/buildAndTest.yml:186]: We can delete this https://github.com/llvm/torch-mlir/blob/51bfe25c898e5f7a09294ccc012cfbf1327614c9/build_tools/build_libtorch.sh#L134 in either this commit or a follow on and switch to src.

**Comment by sjain-stanford** [comment]
> The arm64 Pytorch src build should work if you remove this one line

Thanks, patched. Let's wait to see if the arm64 pytorch source workflow goes through with the fix. If there are more errors, I can revert to pytorch binary and land it for now (to avoid cache evictions the longer this is open). I'll wait for an "all green" CI before landing, but if this looks good otherwise, please feel free to ✅  this.

**Review by powderluv — APPROVED** [review]
LGTM. Feel free to switch arm64 pytorch source in a follow on.

**Comment by powderluv** [comment]
nicely done. The silly cache gets generated again when it merges so we got to wait for it again

**Comment by sjain-stanford** [comment]
> The silly cache gets generated again when it merges so we got to wait for it again

Ah I was wondering why it didn't restore from cache after landing because the keys didn't change. Good to know this is normal. Maybe it treats GHA runs on PRs differently than runs on push to main. Oh well...

**Comment by sjain-stanford** [comment]
... and thank you for the help in reviewing it!

**Pull request merged**
